### ISSUE TITLE
Storage mocks don't appear to correctly model concurrency behaviours of real service fabric objects

### ISSUE THREAD

**Issue by Scythen** [loekd/ServiceFabric.Mocks#27]
Some of my unit tests appear to behave significantly differently when run against mocks vs real storage when more than one thread is accessing the reliable storage.
It would be great if the mocks also mocked the synchronization model of the real objects.

**Comment by loekd** [comment]
If you're talking about transaction isolation, you're right. [This ](https://github.com/loekd/ServiceFabric.Mocks/issues/10)issue is related.
I haven't come up with a way to deal with that yet. Community help is welcome. Do you have time to create a PR?

**Comment by ralphcu** [comment]
I started writing my own Service Fabric mocks for a reliable service I was working on. I have a more complete implementation of the reliable collections.

Things it supports: OnDictionaryChanged, OnTransactionChanged, CommitAsync(), and Abort().
Things it doesn't yet support: IReliableQueue isn't blocking. IReliableDictionary allows dirty read. Neither of these were required for the level of testing I was doing. I have an idea to fix the dirty read problem, but have not put any thought into blocking on the IReliableQueue.

Maybe I could work to add the concepts I implemented into this project so I can switch to use these mocks for my tests and avoid a lot of future maintenance on implementation.

**Comment by ralphcu** [comment]
I played around this weekend writing a transaction aware implementation of the reliable collections. In so doing, I tried one more time to simulate the locking and blocking characteristics of the SF reliable collections. I now have a commit that has these enhancements and passes (with some modifications) the unit tests. More unit tests are needed though. After I write some tests for my lock and lock manager classes, I'll see if I can get a PR up.
